runtime/src/accounts_db.rs

```diff
@@ -753,6 +776,66 @@ impl AccountsDB {
         (reclaims, dead_keys)
     }
 
+    // Construct a vec of the keys needed for cleaning from uncleaned_keys and
+    // zero_lamport_keys sets.
+    fn get_clean_keys(
+        &self,
+        max_clean_root: Option<Slot>,
+        timings: &mut CleanKeyTimings,
+    ) -> Vec<Pubkey> {
+        let mut zero_lamport_key = Measure::start("zero_lamport_key");
+        let pubkeys = self.accounts_index.zero_lamport_accounts.clone();
+        timings.zero_lamport_count = pubkeys.len() as u64;
+        zero_lamport_key.stop();
+        timings.zero_lamport_key += zero_lamport_key.as_us();
+
+        let mut key_create = Measure::start("key_create");
+        let delta_keys: Vec<_> = {
+            let max_slot = max_clean_root.unwrap_or_else(|| self.accounts_index.max_root());
+            let slots: Vec<Slot> = self
+                .uncleaned_keys
+                .iter()
+                .filter_map(|entry| {
+                    let slot = entry.key();
+                    if *slot <= max_slot {
+                        Some(*slot)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            slots
+                .into_iter()
+                .filter_map(|slot| {
+                    let keys = self.uncleaned_keys.remove(&slot).unwrap().1;
```

Comment:
For some future proofing, I think we should move the `unwrap()`call to inside the `if self.accounts_index.is_root(slot)` call below. 

This is just in case in the future we decide to move the handling of `Bank::drop() -> purge_slots()` logic into a separate thread than the AccountsBackgroundService (if that starts getting backed up).

And then we can add the following comment for why the unwrap is safe:

"This function is called from `clean_accounts()`, so no roots from the `accounts_index` can be purged in between when we filtered for roots above and now. Thus because `purge_slots()` only removes unrooted slots, and `clean_accounts()` is the only logic that purges rooted slots, then this slot must still exist in the map."